File: 518-CoinChangeIi/518-CoinChangeIi.java
// Last updated: 2/27/2026, 5:03:02 PM
class Solution {
    public int change(int amount, int[] coins) {
        int[][] dp = new int[coins.length][amount + 1];

        for(int[] arr : dp) Arrays.fill(arr,-1);

        return func(coins, coins.length - 1, amount,dp);
    }

    public int func(int[] coins, int index, int amount, int[][] dp){
        if(amount < 0) return 0;
        if(amount == 0) return 1;

        if(index == 0){
            if(amount % coins[index] == 0) return 1;
            else return 0;
        }

        if(dp[index][amount] != -1) return dp[index][amount];
            
        dp[index][amount] = func(coins, index - 1, amount,dp) + func(coins,index, amount - coins[index],dp);

        return dp[index][amount];

    }
}

File: 518-CoinChangeIi/518-CoinChangeIi.py
# Last updated: 7/12/2026, 6:19:23 PM
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        memo = [[-1] * (amount + 1) for _ in range(n)]

        def dfs(i, a):
            if a == 0: return 1
            if i < 0: return 0
            if a < 0: return 0
            
            if memo[i][a] != -1: return memo[i][a]

            pick = dfs(i, a - coins[i])
            notPick = dfs(i - 1, a)

            memo[i][a] = pick + notPick

            return memo[i][a]


        return dfs(n - 1, amount)

       